### PR TITLE
New Resource & Data Source: azurerm_databox_job

### LINKED ISSUE #6101: Support for new RP DataBoxJob
Description
There is a request here to support data box job, which lets you send terabytes of local data into Azure in a quick, inexpensive, and reliable way. After discussion between our PM and hashicorp owners, we will continue to work on the implementation.
Overview: https://docs.microsoft.com/en-us/azure/databox/data-box-overview
Rest api spec: https://github.com/Azure/azure-rest-api-specs/blob/master/specification/databox/resource-manager/Microsoft.DataBox/stable/2019-09-01/databox.json
PS: https://docs.microsoft.com/en-us/powershell/module/az.databox/new-azdataboxjob?view=azps-3.4.0

Affected resource provider
Microsoft.DataBox

Potential Terraform Configuration
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "example" {
  name     = "${var.prefix}-resources"
  location = var.location
}

resource "azurerm_storage_account" "example" {
  name                     = "${var.prefix}storageaccount"
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location

  account_kind             = "StorageV2"
  account_tier             = "Standard"
  account_replication_type = "RAGRS"
}

resource "azurerm_databox_job" "example" {
  name                = "${var.prefix}-databoxjob"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location

  contact_details {
    name         = "DataBoxJobTester"
    emails       = ["some.user@example.com"]
    phone_number = "+112345678912"
  }

  destination_account {
    type               = "StorageAccount"
    storage_account_id = azurerm_storage_account.example.id
  }

  preferred_shipment_type = "MicrosoftManaged"

  shipping_address {
    city              = "San Francisco"
    country           = "US"
    postal_code       = "94107"
    state_or_province = "CA"
    street_address_1  = "16 TOWNSEND ST"
  }

  sku_name = "DataBox"
}
```

### PR DESCRIPTION
**Pull request by neil-yechenwei**
fixes #6101

### CODE REVIEW

File: azurerm/internal/services/databox/validate/databox_job.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	return warnings, errors
+}
+
+func DataBoxJobPhoneExtension(v interface{}, k string) (warnings []string, errors []error) {
+	value := v.(string)
+
+	if !regexp.MustCompile(`^[\d]{0,4}$`).MatchString(value) {
+		errors = append(errors, fmt.Errorf("%q must have maximum 4 characters in length and can only contain numbers", k))
```
Comment (WodansSon) [inline azurerm/internal/services/databox/validate/databox_job.go]:
```suggestion
		errors = append(errors, fmt.Errorf("%q must contain only numeric values with a maximum length of 4 digits", k))
```

File: azurerm/internal/services/databox/validate/databox_job.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+	return warnings, errors
+}
+
+func DataBoxJobDiskPassKey(v interface{}, k string) (warnings []string, errors []error) {
+	value := v.(string)
+
+	if len(value) < 12 || len(value) > 32 {
+		errors = append(errors, fmt.Errorf(
+			"%q must be between 12 and 32 characters: %q", k, value))
```
Comment (WodansSon) [inline azurerm/internal/services/databox/validate/databox_job.go]:
```suggestion
			"%q must be between 12 and 32 characters in length, received: %q", k, value))
```

File: azurerm/internal/services/databox/validate/databox_job.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	if !regexp.MustCompile(`^.*[\d]+.*[^a-zA-Z0-9 ]+.*|.*[^a-zA-Z0-9 ]+.*[\d]+.*$`).MatchString(value) {
+		errors = append(errors, fmt.Errorf("%q must be alphanumeric, contain at least one special character and atleast one number", k))
```
Comment (WodansSon) [inline azurerm/internal/services/databox/validate/databox_job.go]:
```suggestion
		errors = append(errors, fmt.Errorf("%q must be alphanumeric, contain at least one special character and at least one number", k))
```

File: website/docs/d/databox_job.html.markdown [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+* `device_password` - The device password for unlocking DataBox Heavy.
+
+* `sku_name` - The sku name.
+
+* `tags` - A mapping of tags to assign to the resource.
+
+---
+
+An `destination_account` block supports the following:
```
Comment (WodansSon) [inline website/docs/d/databox_job.html.markdown]:
```suggestion
A `destination_account` block exports the following:
```

File: azurerm/internal/services/databox/data_source_databox_job.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	resp, err := client.Get(ctx, resourceGroup, name, "Details")
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("Error: DataBox Job (DataBox Job Name %q / Resource Group %q) was not found", name, resourceGroup)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/data_source_databox_job.go]: Updated

File: azurerm/internal/services/databox/data_source_databox_job.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		}
+		return fmt.Errorf("Error reading DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/data_source_databox_job.go]: Updated

File: azurerm/internal/services/databox/data_source_databox_job.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+	d.Set("sku_name", resp.Sku.Name)
+	if props := resp.JobProperties; props != nil {
+		if details := props.Details; details != nil {
+			if v, ok := details.AsJobDetailsType(); ok && v != nil {
+				d.Set("device_password", v.DevicePassword)
+
+				if err := d.Set("destination_account", flattenArmDataBoxJobDestinationAccount(v.DestinationAccountDetails)); err != nil {
+					return fmt.Errorf("Error setting `destination_account`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/data_source_databox_job.go]: Updated

File: azurerm/internal/services/databox/data_source_databox_job.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+				}
+			} else if v, ok := details.AsDiskJobDetails(); ok && v != nil {
+				if err := d.Set("destination_account", flattenArmDataBoxJobDestinationAccount(v.DestinationAccountDetails)); err != nil {
+					return fmt.Errorf("Error setting `destination_account`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/data_source_databox_job.go]: Updated

File: azurerm/internal/services/databox/data_source_databox_job.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+				}
+			} else if v, ok := details.AsHeavyJobDetails(); ok && v != nil {
+				d.Set("device_password", v.DevicePassword)
+
+				if err := d.Set("destination_account", flattenArmDataBoxJobDestinationAccount(v.DestinationAccountDetails)); err != nil {
+					return fmt.Errorf("Error setting `destination_account`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/data_source_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 10]
```diff
@@ -0,0 +1,1032 @@
...
+								Type:         schema.TypeString,
+								ValidateFunc: validate.DataBoxJobEmail,
+							},
+						},
+						"phone_number": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.DataBoxJobPhoneNumber,
+						},
+						"mobile": {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 11]
```diff
@@ -0,0 +1,1032 @@
...
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"contact_details": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 12]
```diff
@@ -0,0 +1,1032 @@
...
+					},
+				},
+			},
+
+			"destination_account": {
+				Type:     schema.TypeSet,
+				Required: true,
+				MaxItems: 10,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 13]
```diff
@@ -0,0 +1,1032 @@
...
+					string(databox.MicrosoftManaged),
+				}, false),
+			},
+
+			"shipping_address": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 14]
```diff
@@ -0,0 +1,1032 @@
...
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.DataBoxJobName,
+			},
+
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"contact_details": {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 15]
```diff
@@ -0,0 +1,1032 @@
...
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validate.DataBoxJobStreetAddress,
+						},
+						"street_address_3": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validate.DataBoxJobStreetAddress,
+						},
+						"postal_code_ext": {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: here seems should be zip_code_extension.
image

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 16]
```diff
@@ -0,0 +1,1032 @@
...
+				},
+				PreferredDataCenterRegion: utils.ExpandStringSlice(datacenterRegionPreference),
+			},
+			ShippingAddress: expandArmDataBoxJobShippingAddress(shippingAddress),
+		}
+	}
+
+	future, err := client.Create(ctx, resourceGroup, name, parameters)
+	if err != nil {
+		return fmt.Errorf("Error creating DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 17]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for creation of DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 18]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name, "Details")
+	if err != nil {
+		return fmt.Errorf("Error retrieving DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 19]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("Cannot read DataBox Job (DataBox Job Name %q / Resource Group %q) ID", name, resourceGroup)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 20]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.Name, "Details")
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] DataBox Job %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 21]
```diff
@@ -0,0 +1,1032 @@
...
+			d.Set("delivery_scheduled_date_time", (*props.DeliveryInfo.ScheduledDateTime).Format(time.RFC3339))
+		}
+		d.Set("delivery_type", props.DeliveryType)
+
+		if details := props.Details; details != nil {
+			if v, ok := details.AsJobDetailsType(); ok && v != nil {
+				d.Set("device_password", v.DevicePassword)
+
+				if err := d.Set("contact_details", flattenArmDataBoxJobContactDetails(v.ContactDetails)); err != nil {
+					return fmt.Errorf("Error setting `contact_details`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 22]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if err := d.Set("destination_account", flattenArmDataBoxJobDestinationAccount(v.DestinationAccountDetails)); err != nil {
+					return fmt.Errorf("Error setting `destination_account`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 23]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if v.Preferences != nil {
+					if v.Preferences.TransportPreferences != nil {
+						if err := d.Set("preferred_shipment_type", v.Preferences.TransportPreferences.PreferredShipmentType); err != nil {
+							return fmt.Errorf("Error setting `preferred_shipment_type`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 24]
```diff
@@ -0,0 +1,1032 @@
...
+						}
+					}
+					if err := d.Set("datacenter_region_preference", utils.FlattenStringSlice(v.Preferences.PreferredDataCenterRegion)); err != nil {
+						return fmt.Errorf("Error setting `datacenter_region_preference`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 25]
```diff
@@ -0,0 +1,1032 @@
...
+					}
+				}
+				if err := d.Set("shipping_address", flattenArmDataBoxJobShippingAddress(v.ShippingAddress)); err != nil {
+					return fmt.Errorf("Error setting `shipping_address`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 26]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+			} else if v, ok := details.AsDiskJobDetails(); ok && v != nil {
+				if err := d.Set("databox_preferred_disk", flattenArmDataBoxJobPreferredDisk(v.PreferredDisks)); err != nil {
+					return fmt.Errorf("Error setting `databox_preferred_disk`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: UPdated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 27]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if err := d.Set("contact_details", flattenArmDataBoxJobContactDetails(v.ContactDetails)); err != nil {
+					return fmt.Errorf("Error setting `contact_details`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 28]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if err := d.Set("destination_account", flattenArmDataBoxJobDestinationAccount(v.DestinationAccountDetails)); err != nil {
+					return fmt.Errorf("Error setting `destination_account`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 29]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if v.Preferences != nil {
+					if v.Preferences.TransportPreferences != nil {
+						if err := d.Set("preferred_shipment_type", v.Preferences.TransportPreferences.PreferredShipmentType); err != nil {
+							return fmt.Errorf("Error setting `preferred_shipment_type`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 30]
```diff
@@ -0,0 +1,1032 @@
...
+						}
+					}
+					if err := d.Set("datacenter_region_preference", utils.FlattenStringSlice(v.Preferences.PreferredDataCenterRegion)); err != nil {
+						return fmt.Errorf("Error setting `datacenter_region_preference`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 31]
```diff
@@ -0,0 +1,1032 @@
...
+					}
+				}
+				if err := d.Set("shipping_address", flattenArmDataBoxJobShippingAddress(v.ShippingAddress)); err != nil {
+					return fmt.Errorf("Error setting `shipping_address`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 32]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+			} else if v, ok := details.AsHeavyJobDetails(); ok && v != nil {
+				d.Set("device_password", v.DevicePassword)
+
+				if err := d.Set("contact_details", flattenArmDataBoxJobContactDetails(v.ContactDetails)); err != nil {
+					return fmt.Errorf("Error setting `contact_details`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 33]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if err := d.Set("destination_account", flattenArmDataBoxJobDestinationAccount(v.DestinationAccountDetails)); err != nil {
+					return fmt.Errorf("Error setting `destination_account`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 34]
```diff
@@ -0,0 +1,1032 @@
...
+				}
+				if v.Preferences != nil {
+					if v.Preferences.TransportPreferences != nil {
+						if err := d.Set("preferred_shipment_type", v.Preferences.TransportPreferences.PreferredShipmentType); err != nil {
+							return fmt.Errorf("Error setting `preferred_shipment_type`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 35]
```diff
@@ -0,0 +1,1032 @@
...
+						}
+					}
+					if err := d.Set("datacenter_region_preference", utils.FlattenStringSlice(v.Preferences.PreferredDataCenterRegion)); err != nil {
+						return fmt.Errorf("Error setting `datacenter_region_preference`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 36]
```diff
@@ -0,0 +1,1032 @@
...
+					}
+				}
+				if err := d.Set("shipping_address", flattenArmDataBoxJobShippingAddress(v.ShippingAddress)); err != nil {
+					return fmt.Errorf("Error setting `shipping_address`: %+v", err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 37]
```diff
@@ -0,0 +1,1032 @@
...
+				ContactDetails:  expandArmDataBoxJobContactDetails(contactDetails),
+				ShippingAddress: expandArmDataBoxJobShippingAddress(shippingAddress),
+			},
+		},
+		Tags: tags.Expand(t),
+	}
+
+	future, err := client.Update(ctx, resourceGroup, name, parameters, "")
+	if err != nil {
+		return fmt.Errorf("Error updating DataBox Job %q (Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 38]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for update of DataBox Job %q (Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 39]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name, "Details")
+	if err != nil {
+		return fmt.Errorf("Error retrieving DataBox Job %q (Resource Group %q): %+v", name, resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 40]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("Cannot read DataBox Job %s (resource group %s) ID", name, resourceGroup)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 41]
```diff
@@ -0,0 +1,1032 @@
...
+		Reason: utils.String("Cancel the order for deleting"),
+	}
+
+	resp, err := client.Cancel(ctx, id.ResourceGroup, id.Name, *reason)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp) {
+			return nil
+		}
+
+		return fmt.Errorf("Error cancelling Order (DataBox Job Name %q / Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 42]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+
+	future, err := client.Delete(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("Error deleting DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 43]
```diff
@@ -0,0 +1,1032 @@
...
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for deleting DataBox Job (DataBox Job Name %q / Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go [hunk 44]
```diff
@@ -0,0 +1,1032 @@
...
+			}
+
+			if scope != "" {
+				resp, err := lockClient.DeleteByScope(ctx, scope, lockName)
+				if err != nil {
+					if utils.ResponseWasNotFound(resp) {
+						return nil
+					}
+
+					return fmt.Errorf("Error deleting Management Lock (Lock Name %q / Scope %q): %+v", lockName, scope, err)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: updated

File: azurerm/internal/services/databox/validate/databox_job.go [hunk 45]
```diff
@@ -0,0 +1,128 @@
...
+
+func DataBoxJobID(i interface{}, k string) (warnings []string, errors []error) {
+	v, ok := i.(string)
+	if !ok {
+		errors = append(errors, fmt.Errorf("expected type of %q to be string", k))
+		return
+	}
+
+	if _, err := parse.DataBoxJobID(v); err != nil {
+		errors = append(errors, fmt.Errorf("Can not parse %q as a resource id: %v", k, err))
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/validate/databox_job.go]: updated

File: azurerm/internal/services/databox/validate/databox_job.go [hunk 46]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	return warnings, errors
+}
+
+func DataBoxJobName(v interface{}, k string) (warnings []string, errors []error) {
+	value := v.(string)
+
+	if !regexp.MustCompile(`^[\da-zA-Z][-\da-zA-Z]{1,22}[\da-zA-Z]$`).MatchString(value) {
+		errors = append(errors, fmt.Errorf("%q must be between 3 and 24 characters in length, and it must begin and end with an alphanumeric and can only contain alphanumeric characters and hyphens", k))
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/validate/databox_job.go]: updated

File: azurerm/internal/services/databox/validate/databox_job.go [hunk 47]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	return warnings, errors
+}
+
+func DataBoxJobPhoneNumber(v interface{}, k string) (warnings []string, errors []error) {
+	value := v.(string)
+
+	if !regexp.MustCompile(`^[+][\d]{2,}$`).MatchString(value) {
+		errors = append(errors, fmt.Errorf("%q must begin with + and may contain only at least 2 numbers", k))
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/validate/databox_job.go]: updated

File: azurerm/internal/services/databox/validate/databox_job.go (hunk 1 above)
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/validate/databox_job.go]: updated

File: azurerm/internal/services/databox/validate/databox_job.go (hunk 2 above)
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/validate/databox_job.go]: updated

File: azurerm/internal/services/databox/validate/databox_job.go (hunk 3 above)
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/validate/databox_job.go]: updated

File: website/docs/d/databox_job.html.markdown (hunk 4 above)
Comment (neil-yechenwei) [reply website/docs/d/databox_job.html.markdown]: updated

File: azurerm/internal/services/databox/resource_arm_databox_job.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply azurerm/internal/services/databox/resource_arm_databox_job.go]: Updated

**Comment by neil-yechenwei** [comment]
@WodansSon , shipping_address is required for api call, so I cannot add test case for "shipping_address = "None"". For other suggestions, I've added test cases and they all passed after tested.

**Comment by katbyte** [comment]
As we cannot easily/safely tests this (ie a known test/other way to ensure a databox doesn't get sent out) i'm going to close it. Once the API/service has a way to flag a request as "test/ignore" we can revive the branch on gh and get this merged

**Comment by WodansSon** [comment]
The service team has now exposed a physical address we can use to test this resource, re-opening PR per @katbyte comment on July 6th...

**Comment by WodansSon** [comment]
We need to pull this one out of the release because the service team has informed us that the shipping mechanisms in Data Box Job and Data Box Edge are different and the test's for Data Box Job will actually ship live data boxes to the test address.

**Comment by WodansSon** [comment]
As the shipping mechanism in this version of the databox does not allow for testing of the resource, we will not support this version of the API in Terraform until the service team can address the issue.
